File: CHTL/CHTLNode/ElementNode.cpp
#include "ElementNode.h"

namespace CHTL {

ElementNode::ElementNode(const std::string& tagName) : tagName(tagName) {}

void ElementNode::accept(Visitor& visitor) {
    visitor.visit(*this);
}

void ElementNode::addChild(std::unique_ptr<BaseNode> child) {
    children.push_back(std::move(child));
}

void ElementNode::addAttribute(const HtmlAttribute& attr) {
    attributes.push_back(attr);
}

std::unique_ptr<BaseNode> ElementNode::clone() const {
    auto new_node = std::make_unique<ElementNode>(this->tagName);
    for (const auto& attr : this->attributes) {
        HtmlAttribute new_attr;
        new_attr.key = attr.key;
        new_attr.value = attr.value;
        new_attr.value_expr = attr.value_expr ? attr.value_expr->clone() : nullptr;
        new_node->attributes.push_back(new_attr);
    }
    for (const auto& child : this->children) {
        new_node->addChild(child->clone());
    }
    return new_node;
}

} // namespace CHTL

File: CHTL/CHTLNode/ElementNode.py
from .BaseNode import BaseNode

class ElementNode(BaseNode):
    """
    Represents an element in CHTL, which corresponds to an HTML tag.
    e.g., div { ... }
    """
    def __init__(self, tag_name: str, children: list[BaseNode] = None, attributes: dict[str, str] = None):
        self.tag_name = tag_name
        self.children = children if children is not None else []
        self.attributes = attributes if attributes is not None else {}

    def to_dict(self) -> dict:
        """Converts the element node to a dictionary."""
        return {
            "type": "ElementNode",
            "tag_name": self.tag_name,
            "attributes": self.attributes,
            "children": [child.to_dict() for child in self.children]
        }

    def __repr__(self) -> str:
        return f"<ElementNode tag='{self.tag_name}' attributes={len(self.attributes)} children={len(self.children)}>"
